### PR TITLE
Add run command for maven/gradle

### PR DESCRIPTION
**Pull request by florianGla**

### CODE REVIEW

File: internal/builder/builder.go [hunk 1]
```diff
@@ -184,12 +184,7 @@ func (i *CIFuzzBuilder) BuildCIFuzzAndDeps() error {
 		return err
 	}
 
-	err = i.CopyVSCodeTasks()
-	if err != nil {
-		return err
-	}
-
-	err = i.CopyLogo()
```
Comment (jochil) [inline internal/builder/builder.go:192]: I would guess we can delete the `CopyLogo` function too, or?

Comment (florianGla) [reply internal/builder/builder.go:192]: Yes, you are right

File: pkg/dependencies/definitions_test.go [hunk 2]
```diff
@@ -9,12 +9,12 @@ import (
 )
 
 func TestResetDefault(t *testing.T) {
-	oldDefault := Default
-	Default = Dependencies{}
+	oldDefault := CMakeDeps
```
Comment (jochil) [inline pkg/dependencies/definitions_test.go:12]: can you also add some assertions for the new `MavenDeps` and `GradleDeps` to this test?

File: pkg/dependencies/deps.go [hunk 3]
```diff
@@ -17,6 +17,9 @@ const (
 	LLVM_SYMBOLIZER Key = "llvm-symbolizer"
 	LLVM_PROFDATA   Key = "llvm-profdata"
 
+	MAVEN  Key = "maven"
```
Comment (jochil) [inline pkg/dependencies/deps.go:20]: for all other cases the key is also the name of the executable... what do you think about using `mvn` here?

**Comment by jochil** [comment]
@florianGla @fmeum could you please outline the changes that are necessary after the jazzer refactoring is done? Maybe we can mark them in the code with some `TODO`s. IMO it would help us later

File: internal/build/gradle/gradle.go [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+	"code-intelligence.com/cifuzz/internal/build"
+	"code-intelligence.com/cifuzz/internal/cmdutils"
+	"code-intelligence.com/cifuzz/pkg/log"
+	"code-intelligence.com/cifuzz/pkg/runfiles"
+	"code-intelligence.com/cifuzz/util/fileutil"
+)
+
+var classpathRegex = regexp.MustCompile("(?m)^Test classpath = (?P<classpath>.*)$")
+
+func FindGradleWrapper(projectDir string) (string, error) {
```
Comment (jochil) [inline internal/build/gradle/gradle.go:23]: In the gradle docs (https://docs.gradle.org/current/userguide/gradle_wrapper.html) it sounds like the Wrapper is recommended but not necessary? Do you think we need a fallback to the `gradle` command if the wrapper is missing?

**Comment by jochil** [comment]
@florianGla can you also update the READMEs of the maven/gradle example projects? At least the neccessary `cifuzz run` command would be helpful

File: internal/build/gradle/gradle.go [hunk 5]
```diff
@@ -0,0 +1,178 @@
...
+	output, err := cmd.Output()
+	if err != nil {
+		// It's expected that gradle might fail due to user configuration,
+		// so we print the error without the stack trace.
+		err = cmdutils.WrapExecError(err, cmd)
+		log.Error(err)
+		return nil, cmdutils.ErrSilent
+	}
+	classpath := classpathRegex.FindStringSubmatch(string(output))
+	deps := strings.Split(string(classpath[1]), ":")
```
Comment (fmeum) [inline internal/build/gradle/gradle.go:153]: The class path separator is `os.PathListSeparator`, which is `;` on Windows.

File: internal/build/gradle/gradle.go [hunk 6]
```diff
@@ -0,0 +1,178 @@
...
+
+	return deps, nil
+}
+
+func (b *Builder) getFuzzTestSeedCorpus(targetClass string) string {
+	path := strings.Split(targetClass, ".")
```
Comment (fmeum) [inline internal/build/gradle/gradle.go:159]: We currently store the seed corpus in `<ClassName>SeedCorpus`, so the last segment has to be modified.

File: internal/build/gradle/gradle.go [hunk 7]
```diff
@@ -0,0 +1,178 @@
...
+func (b *Builder) getGradleCommand() (string, error) {
+	wrapper, err := FindGradleWrapper(b.ProjectDir)
+	if err != nil {
+		return "", err
+	}
+	if wrapper != "" {
+		return wrapper, nil
+	}
+
+	return "gradle", nil
```
Comment (fmeum) [inline internal/build/gradle/gradle.go:177]: Shouldn't we use the runfiles finder here?

File: internal/build/maven/maven.go [hunk 8]
```diff
@@ -0,0 +1,153 @@
...
+	}
+
+	bytes, err := os.ReadFile(outputPath)
+	if err != nil {
+		return nil, errors.WithStack(err)
+	}
+
+	deps := strings.Split(string(bytes), ":")
+	// Append dependencies which are not listed by "mvn dependency:build-classpath"
+	localDeps := []string{"target/classes", "target/test-classes", "resource", "test-resource"}
```
Comment (fmeum) [inline internal/build/maven/maven.go:140]: These should use `filepath.Join`. Are the third and fourth entries correct as they are or do they also lie under `target`?

File: internal/build/maven/maven.go [hunk 9]
```diff
@@ -0,0 +1,153 @@
...
+	for _, dep := range localDeps {
+		deps = append(deps, filepath.Join(b.ProjectDir, dep))
+	}
+
+	return deps, nil
+}
+
+func (b *Builder) getFuzzTestSeedCorpus(targetClass string) string {
+	path := strings.Split(targetClass, ".")
```
Comment (fmeum) [inline internal/build/maven/maven.go:149]: See the commend on Gradle. Could we extract this method?

File: internal/build/maven/maven.go (hunk 8 above)
Comment (florianGla) [reply internal/build/maven/maven.go:140]: You are right, I think they also lie under target.

**Comment by fmeum** [comment]
> @florianGla @fmeum could you please outline the changes that are necessary after the jazzer refactoring is done? Maybe we can mark them in the code with some `TODO`s. IMO it would help us later

There will be smaller changes required in multiple places, but nothing that changes the overall architecture. I would rather not add comments describing a version that hasn't been released yet - we may make further changes to it and the current approach will continue to work.

File: integration-tests/gradle/gradle_test.go [hunk 10]
```diff
@@ -31,7 +38,18 @@ func TestIntegration_Gradle(t *testing.T) {
...
 
 	// Execute the create command
-	outputPath := filepath.Join("src", "MyClassFuzzTest.java")
+	testDir := filepath.Join(
+		"src",
+		"test",
+		"java",
+		"com",
+		"github",
+		"CodeIntelligenceTesting",
```
Comment (fmeum) [inline integration-tests/gradle/gradle_test.go:47]: Package names should always be in lower case. You can use `com/example` or `com/github/code_intelligence`.

File: integration-tests/maven/maven_test.go [hunk 11]
```diff
@@ -1,7 +1,14 @@
 package maven
 
 import (
+	"bufio"
+	"code-intelligence.com/cifuzz/pkg/parser/libfuzzer/stacktrace"
```
Comment (fmeum) [inline integration-tests/maven/maven_test.go:5]: This should be reformatted with `make fmt`. You can install the push hook in `githooks` to fix this automatically.

File: integration-tests/gradle/gradle_test.go (hunk 10 above)
Comment (florianGla) [reply integration-tests/gradle/gradle_test.go:47]: Currently our gradle fuzz test class is located under cifuzz/examples/gradle/src/main/java/com/github/CodeIntelligenceTesting/cifuzz. Shall we change this path and the package name accordingly?

Comment (fmeum) [reply integration-tests/gradle/gradle_test.go:47]: Yes, we should move it to a more common package.

**Comment by jochil** [comment]
> > @florianGla @fmeum could you please outline the changes that are necessary after the jazzer refactoring is done? Maybe we can mark them in the code with some `TODO`s. IMO it would help us later
> 
> There will be smaller changes required in multiple places, but nothing that changes the overall architecture. I would rather not add comments describing a version that hasn't been released yet - we may make further changes to it and the current approach will continue to work.

IIRC the current approach requires the jazzer driver to be installed manually. Or Am I wrong @florianGla?
It will not be shipped with this change... so from user perspective this is not a working change

**Comment by florianGla** [comment]
> IIRC the current approach requires the jazzer driver to be installed manually. Or Am I wrong @florianGla? It will not be shipped with this change... so from user perspective this is not a working change

You are right, I think we should wait for the jazzer standalone package before merging.

File: internal/build/build.go [hunk 12]
```diff
@@ -67,3 +68,11 @@ func CommonBuildEnv() ([]string, error) {
 
 	return env, nil
 }
+
+func JazzerSeedCorpus(targetClass string, projectDir string) string {
+	seedCorpus := targetClass + "SeedCorpus"
```
Comment (jochil) [inline internal/build/build.go:73]: I think the Suffix is `Inputs` instead of `SeedCorpus`... at least after the jazzer changes from @fmeum

Comment (MarkusZoppelt) [reply internal/build/build.go]: Changed!

**Comment by fmeum** [comment]
Based on the log output it looks like our fancy printer may have erased some parts of Jazzer's exception:
```
2022-10-27T13:07:48.2765717Z 
2022-10-27T13:07:48.2765961Z remote code execution depending on available classes on the classpath.eHigh: Remote Code Execution
2022-10-27T13:07:48.2766406Z Unrestricted class loading based on externally controlled data may allow
2022-10-27T13:07:48.2766788Z remote code execution depending on available classes on the classpath.
2022-10-27T13:07:48.2767118Z 	at jaz.Zer.<clinit>(Zer.java:54)
2022-10-27T13:07:48.2767447Z 	at java.base/java.lang.Class.forName0(Native Method)
2022-10-27T13:07:48.2767811Z 	at java.base/java.lang.Class.forName(Class.java:315)
2022-10-27T13:07:48.2768191Z 	at com.example.ExploreMe.exploreMe(ExploreMe.java:13)
2022-10-27T13:07:48.2768606Z 	at com.example.FuzzTestCase.myFuzzTest(FuzzTestCase.java:19)
```

@adombeck Do you see how this could happen?

File: pkg/runfiles/finder.go [hunk 13]
```diff
@@ -55,6 +48,12 @@ func (f RunfilesFinderImpl) LLVMSymbolizerPath() (string, error) {
 	return path, errors.WithStack(err)
 }
 
+func (f RunfilesFinderImpl) JavaPath() (string, error) {
```
Comment (fmeum) [inline pkg/runfiles/finder.go:51]: Is this even used?

File: pkg/runner/jazzer/jazzer_runner.go [hunk 14]
```diff
@@ -63,16 +63,18 @@ func (r *Runner) Run(ctx context.Context) error {
...
-	jazzerPath, err := runfiles.Finder.JazzerPath()
-	if err != nil {
-		return err
-	}
-	agentPath, err := runfiles.Finder.JazzerAgentDeployJarPath()
+	javaHome, err := runfiles.Finder.JavaHomePath()
 	if err != nil {
 		return err
 	}
+	javaBin := filepath.Join(javaHome, "bin", "java")
```
Comment (fmeum) [inline pkg/runner/jazzer/jazzer_runner.go:70]: We have to append `.exe` on Windows.

File: pkg/runfiles/finder.go (hunk 13 above)
Comment (florianGla) [reply pkg/runfiles/finder.go:51]: No, I think were are looking at an old version

Comment (florianGla) [reply pkg/runfiles/finder.go:51]: You are right, I forgot to delete old functions

File: internal/build/maven/maven.go [hunk 15]
```diff
@@ -0,0 +1,148 @@
...
+		} else {
+			// Use one thread per cpu core
+			flags = append(flags, "1C")
+		}
+	}
+
+	args := append(flags, "test-compile")
+
+	cmd := exec.Command("mvn", args...)
+	// Redirect the command's stdout to stderr to only have
```
Comment (fmeum) [inline internal/build/maven/maven.go:74]: See above

File: internal/build/maven/maven.go [hunk 16]
```diff
@@ -0,0 +1,148 @@
...
+	}
+
+	bytes, err := os.ReadFile(outputPath)
+	if err != nil {
+		return nil, errors.WithStack(err)
+	}
+
+	deps := strings.Split(string(bytes), string(os.PathListSeparator))
+	// Append local dependencies which are not listed by "mvn dependency:build-classpath"
+	localDeps := []string{"classes", "test-classes", "resource", "test-resource"}
```
Comment (fmeum) [inline internal/build/maven/maven.go:142]: We should add a comment stating that these directories are configurable and we should really read them from the resolved POM.

**Review by fmeum — APPROVED** [review]
Looks pretty good! The remaining comments are all just minor things, so I'm going to approve right away.

File: integration-tests/gradle/gradle_test.go [hunk 17]
```diff
@@ -49,5 +66,108 @@ func TestIntegration_Gradle(t *testing.T) {
...
+	cifuzzRunner.Run(t, &shared.RunOptions{
+		ExpectedOutputs: []*regexp.Regexp{expectedOutputExp},
+	})
+
+	// Check that the findings command lists the finding
+	findings = shared.GetFindings(t, cifuzz, projectDir)
+	require.Len(t, findings, 1)
+	require.Contains(t, findings[0].Details, "Remote Code Execution")
+
+	//// TODO: This check currently fails on macOS because there
```
[the remark this answers is not in the document]
Comment (MarkusZoppelt) [reply integration-tests/gradle/gradle_test.go]: Fixed, thanks!

File: internal/build/maven/maven.go (hunk 16 above)
Comment (MarkusZoppelt) [reply internal/build/maven/maven.go:142]: Added a comment and opened a ticket for a different PR.

**Pull request merged**